### PR TITLE
arbstudio upgrade

### PR DESCRIPTION
**Pull request by nulinspiratie**
Requires merging PR https://github.com/nulinspiratie/Qcodes/pull/76

Changes:
- Change minimum sine pulse frequency from 1 MHz to 100 Hz
- Add parameter `force_upload_waveform` (default False), which when False does not upload waveforms if they are identical to the ones already uploaded
- Performance improvement to pulse sequence targeting. The old method was very slow for pulse sequences with many pulses.
- Incorporate arbstudio channels  (PR https://github.com/nulinspiratie/Qcodes/pull/76)
- Proper sine pulse generation

### CODE REVIEW

**Review by RostyslavSavytskyy — CHANGES_REQUESTED** [review]

**Review by RostyslavSavytskyy — CHANGES_REQUESTED** [review]

File: silq/instrument_interfaces/lecroy/ArbStudio1104_interface.py [hunk 1]
```diff
@@ -51,54 +54,84 @@ def __init__(self, instrument_name, **kwargs):
...
                                name='trig_in', input_trigger=True),
             'trig_out': Channel(instrument_name=self.instrument_name(),
-                                name='trig_out', output_TTL=(0, 3.3))}
+                                name='trig_out', output_TTL=(0, 3.3))
+        }
         # TODO check Arbstudio output TTL high voltage
 
         self.pulse_implementations = [SinePulseImplementation(
-            pulse_requirements=[('frequency', {'min': 1e6, 'max': 125e6})]),
+            pulse_requirements=[('frequency', {'min': 1e2, 'max': 125e6})]),
```
[the remark this answers is not in the document]
Comment (nulinspiratie) [reply silq/instrument_interfaces/lecroy/ArbStudio1104_interface.py:70]: Agreed, fixed

File: silq/instrument_interfaces/lecroy/ArbStudio1104_interface.py [hunk 2]
```diff
@@ -51,54 +54,84 @@ def __init__(self, instrument_name, **kwargs):
...
-
-        self.add_parameter('active_channels', get_cmd=self._get_active_channels)
-
-    def _get_active_channels(self) -> List[str]:
+        self.trigger_in_duration = Parameter(
+            set_cmd=None,
+            unit='s',
+            initial_value=100e-9,
+            docstring="Required duration of a trigger pulse, to be requested "
+                      "from the pulse that triggers the Arbstudio"
```
[the remark this answers is not in the document]
Comment (nulinspiratie) [reply silq/instrument_interfaces/lecroy/ArbStudio1104_interface.py:81]: Replaced "pulse" by "interface", it should be clear now. If not, please let me know a better description

File: silq/instrument_interfaces/lecroy/ArbStudio1104_interface.py [hunk 3]
```diff
@@ -51,54 +54,84 @@ def __init__(self, instrument_name, **kwargs):
...
-                                     'have full waveforms, to ensure that the '
-                                     'waveform is finished before the next '
-                                     'trigger arrives. This does not count for '
-                                     'pulses such as DCPulse, which only have '
-                                     'four points')
-
-        self.add_parameter('active_channels', get_cmd=self._get_active_channels)
-
-    def _get_active_channels(self) -> List[str]:
+        self.trigger_in_duration = Parameter(
```
[the remark this answers is not in the document]
Comment (nulinspiratie) [reply silq/instrument_interfaces/lecroy/ArbStudio1104_interface.py:76]: yes we'll discuss

File: silq/instrument_interfaces/lecroy/ArbStudio1104_interface.py [hunk 4]
```diff
@@ -51,54 +54,84 @@ def __init__(self, instrument_name, **kwargs):
...
+        )
+        self.pulse_final_delay = Parameter(
+            set_cmd=None,
+            unit='s', initial_value=1e-6,
+            docstring='Final delay up to the end of pulses that '
+                      'have full waveforms, to ensure that the '
+                      'waveform is finished before the next '
+                      'trigger arrives. This does not count for '
+                      'pulses such as DCPulse, which only have '
+                      'four points'
```
[the remark this answers is not in the document]
Comment (nulinspiratie) [reply silq/instrument_interfaces/lecroy/ArbStudio1104_interface.py:91]: Yep good point, I have modified the docstring. Have a look if it's clearer now

File: silq/instrument_interfaces/lecroy/ArbStudio1104_interface.py [hunk 5]
```diff
@@ -124,42 +157,42 @@ def get_additional_pulses(self, connections) -> List[Pulse]:
...
+            for pulse in sorted(pulses, key=lambda p: p.t_start):
+                if pulse.t_start > t:
+                    # Next pulse starts some time after the current point in time.
+                    # Add DC pulse at amplitude zero to bridge the gap
                     dc_pulse = self.get_pulse_implementation(
-                        DCPulse(t_start=t, t_stop=self.pulse_sequence.duration,
-                                amplitude=0, connection=connection))
+                        DCPulse(t_start=t, t_stop=pulse.t_start, amplitude=0, connection=connection)
+                    )
                     self.pulse_sequence.add(dc_pulse)
```
[the remark this answers is not in the document]
Comment (nulinspiratie) [reply silq/instrument_interfaces/lecroy/ArbStudio1104_interface.py:169]: I think this part of the code works pretty much the same way as before, just more efficient. DC pulses at 0V were always added to bridge gaps between pulses

File: silq/instrument_interfaces/lecroy/ArbStudio1104_interface.py [hunk 6]
```diff
@@ -177,45 +210,36 @@ def setup(self, **kwargs):
...
-                                              pulse_class=SinePulse):
+            if self.pulse_sequence.get_pulses(
+                    output_channel=ch, pulse_class=SinePulse
+            ):
                 # TODO better check for when to use burst or stepped mode
-                self.instrument.parameters[f'{ch}_trigger_mode']('burst')
+                channel.trigger_mode = 'burst'
             else:
-                self.instrument.parameters[f'{ch}_trigger_mode']('stepped')
+                channel.trigger_mode = 'stepped'
```
[the remark this answers is not in the document]
Comment (nulinspiratie) [reply silq/instrument_interfaces/lecroy/ArbStudio1104_interface.py:236]: Not really, I don't think burst mode would work for our purposes

File: silq/instrument_interfaces/lecroy/ArbStudio1104_interface.py [hunk 7]
```diff
@@ -331,42 +414,46 @@ def implement(self, sampling_rates, input_pulse_sequence, **kwargs):
...
-            channels = [self.connection.output['channel'].name]
+        if isinstance(self.pulse.connection, SingleConnection):
+            channels = [self.pulse.connection.output['channel'].name]
         else:
-            raise Exception(f"No implementation for connection {self.connection}")
+            raise Exception(f"No implementation for connection {self.pulse.connection}")
 
-        assert self.frequency < min(sampling_rates[ch] for ch in channels)/2, \
-            'Sine frequency is higher than the Nyquist limit for channels {channels}'
+        assert self.pulse.frequency > 0, "Pulse frequency must be larger than zero."
```
[the remark this answers is not in the document]
Comment (nulinspiratie) [reply silq/instrument_interfaces/lecroy/ArbStudio1104_interface.py:428]: fixed

[the remark this answers is not in the document]
Comment (nulinspiratie) [reply silq/instrument_interfaces/lecroy/ArbStudio1104_interface.py:428]: at least now abs(frequency) must be larger than 100. Else the implementation fails.

File: silq/instrument_interfaces/lecroy/ArbStudio1104_interface.py [hunk 8]
```diff
@@ -331,42 +414,46 @@ def implement(self, sampling_rates, input_pulse_sequence, **kwargs):
...
+            if points_per_period > self.max_waveform_points:
+                raise RuntimeError(
+                    f"Sine waveform points {points_per_period} is above maximum "
+                    f"{self.max_waveform_points}. Could not segment sine waveform"
+                )
+            elif points_per_period > 1000:
+                # Temporarily modify pulse frequency to ensure waveforms have full period
+                original_frequency = self.pulse.frequency
+                modified_frequency = sample_rate / points_per_period
+                self.pulse.frequency = modified_frequency
```
[the remark this answers is not in the document]
Comment (nulinspiratie) [reply silq/instrument_interfaces/lecroy/ArbStudio1104_interface.py:444]: Yeah I think so, the frequency would be a tiny bit off, so I temporarily modify it to ensure the waveform has a full period.
The reason for only doing this at low-ish frequencies is that above this value, the change in frequency due to just a single point may be significant. Added a comment to clarify this

File: silq/instrument_interfaces/lecroy/ArbStudio1104_interface.py [hunk 9]
```diff
@@ -331,42 +414,46 @@ def implement(self, sampling_rates, input_pulse_sequence, **kwargs):
...
+
+                t_list = self.pulse.t_start + np.arange(points_per_period) / sample_rate
+                voltages = self.pulse.get_voltage(t_list)
+
+                self.pulse.frequency = original_frequency
+            else:
+                periods = 50000 // points_per_period
+                waveform_points = int(periods * points_per_period)
+                t_list = self.pulse.t_start + np.arange(waveform_points) / sample_rate
+                voltages = self.pulse.get_voltage(t_list)
```
[the remark this answers is not in the document]
Comment (nulinspiratie) [reply silq/instrument_interfaces/lecroy/ArbStudio1104_interface.py:454]: What was the error? I added a comment to clarify

File: silq/instrument_interfaces/lecroy/ArbStudio1104_interface.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (RostyslavSavytskyy) [reply silq/instrument_interfaces/lecroy/ArbStudio1104_interface.py]: good

File: silq/instrument_interfaces/lecroy/ArbStudio1104_interface.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (RostyslavSavytskyy) [reply silq/instrument_interfaces/lecroy/ArbStudio1104_interface.py:170]: ok

File: silq/instrument_interfaces/lecroy/ArbStudio1104_interface.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (RostyslavSavytskyy) [reply silq/instrument_interfaces/lecroy/ArbStudio1104_interface.py]: great

File: silq/instrument_interfaces/lecroy/ArbStudio1104_interface.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (RostyslavSavytskyy) [reply silq/instrument_interfaces/lecroy/ArbStudio1104_interface.py:463]: I already don't remember explicitly, can test it later

**Review by RostyslavSavytskyy — APPROVED** [review]

**Comment by RostyslavSavytskyy** [comment]
Looks good to me

**Pull request merged**